### PR TITLE
Use the latest autorest libraries for type generation

### PR DESCRIPTION
**Pull request by anthony-c-martin**
This PR contains the code changes to switch from C# type generation to TypeScript.

The motivation is that the library we're using (https://github.com/Azure/autorest.common) is not being actively maintained, with modern extensions being authored using https://github.com/Azure/autorest/tree/master/packages/extensions/modelerfour

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -50,4 +49,48 @@ jobs:
       - name: Upload Code Coverage
         uses: codecov/codecov-action@v1.5.0
         with:
-          flags: dotnet
\ No newline at end of file
+          flags: dotnet
+
+  build-generator:
```
Comment (majastrz) [inline .github/workflows/build.yml:54]: >build [](http://example.com/codeflow?start=2&length=5)

do we need official build changes?

Comment (anthony-c-martin) [reply .github/workflows/build.yml:54]: I don't believe so - the official build just triggers the nupkg build (which is unchanged)

**Comment by majastrz** [comment]
Does running the new generator on the same commit as the old one produce an empty diff or a diff with "good" changes?

File: .github/workflows/build.yml (hunk 1 above)
Comment (majastrz) [reply .github/workflows/build.yml:54]: Worth building all the things regardless? (I'm good either way.)

**Review by majastrz — APPROVED** [review]
This is great!

File: src/autorest.bicep/src/type-generator.ts [hunk 2]
```diff
@@ -0,0 +1,451 @@
...
+
+    if (putProperty && putProperty.required) {
+      flags |= ObjectPropertyFlags.Required;
+    }
+
+    if (!putProperty || putProperty.readOnly) {
+      flags |= ObjectPropertyFlags.ReadOnly;
+    }
+
+    if (!getProperty) {
```
Comment (shenglol) [inline src/autorest.bicep/src/type-generator.ts:303]: Should we also handle `"secret": true` and `"x-ms-mutability": ["create", "update"]`?

File: src/autorest.bicep/src/resources.ts [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+
+  return path;
+}
+
+export function parseNameSchema<T>(descriptor: ResourceDescriptor, request: HttpRequest, parameters: Parameter[], parseType: (schema: Schema) => T, createConstantName: (name: string) => T) {
+  const path = getNormalizedMethodPath(request.path);
+
+  const finalProvidersMatch = path.match(parentScopePrefix)?.last;
+  if (!finalProvidersMatch) {
+    return { success: false, failureReason: `Unable to locate "/providers/" segment` };
```
Comment (shenglol) [inline src/autorest.bicep/src/resources.ts:93]: Might be worth extracting two functions`failure(failureReason: string)` and `success(name: string)`?

File: src/autorest.bicep/tsconfig.json [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    "alwaysStrict": true,
+    "forceConsistentCasingInFileNames": false,
+    "module": "commonjs",
+    "noEmitOnError": true,
+    "noImplicitAny": true,
+    "noImplicitReturns": true,
+    "noImplicitThis": true,
+    "sourceMap": true,
+    "declarationMap": true,
+    "strict": true,
```
Comment (shenglol) [inline src/autorest.bicep/tsconfig.json:13]: Btw `"strict": true` implies `"alwaysStrict": true`, `"noImplictAny": true` and `"noImplictThis": true` so they could safely be removed (it doesn't hurt to keep them though).

File: .github/workflows/build.yml (hunk 1 above)
Comment (anthony-c-martin) [reply .github/workflows/build.yml:54]: I'm fine with the official build just building the releasable assets for now - CI on the main branch and on PRs should catch any regressions in the generator code.

**Comment by anthony-c-martin** [comment]
> Does running the new generator on the same commit as the old one produce an empty diff or a diff with "good" changes?

Yes - I used the diffs heavily to progressively work towards this implementation. I spent some time going through the final product and coudln't find anything major other than naming and formatting differences.

File: src/autorest.bicep/src/type-generator.ts (hunk 2 above)
Comment (anthony-c-martin) [reply src/autorest.bicep/src/type-generator.ts:303]: Good point. I'll raise an issue for this, as the intention with this PR is to reach parity with the current generator.

Comment (anthony-c-martin) [reply src/autorest.bicep/src/type-generator.ts:303]: Raised #240

Comment (shenglol) [reply src/autorest.bicep/src/type-generator.ts:303]: Sounds good.

**Pull request merged**